File: LeetCode/ConstructAlgorithm/Golang/Leetcode1775-equal-sum-arrays-with-minimum-number-of-operations.go
package sbgo

func minOperations(nums1 []int, nums2 []int) int {
  if len(nums1) * 6 < len(nums2) || len(nums2) * 6 < len(nums1) {
    return -1
  }
  getsum := func(nums []int) int {
    res := 0
    for _, x := range nums {
      res += x
    }
    return res
  }
  d := getsum(nums1) - getsum(nums2)
  if d < 0 {
    nums1, nums2 = nums2, nums1
    d = -d
  }
  cnt := make([]int, 6)
  for _, x := range nums1 {
    cnt[x - 1]++
  }
  for _, x := range nums2 {
    cnt[6 -x]++
  }
  ans := 0
  for i := 5;i > 0;i-- {
    if d - cnt[i] * i <= 0 {
      return ans + (d + i - 1) / i 
    }
    d -= cnt[i] * i
    ans += cnt[i]
  }
  return ans    
}

File: LeetCode/ConstructAlgorithm/Rust/Leetcode1775-equal-sum-arrays-with-minimum-number-of-operations.rs
use std::mem::swap;
impl Solution {
  pub fn min_operations(mut nums1: Vec<i32>, mut nums2: Vec<i32>) -> i32 {
    if 6 * nums1.len() < nums2.len() || 6 * nums2.len() < nums1.len() {
      return -1;
    }
    let mut d: i32 = nums1.iter().sum::<i32>() - nums2.iter().sum::<i32>();
    if d < 0 {
      swap(&mut nums1, &mut nums2);
      d = -d;
    }
    let (mut cnt, mut ans) = (vec![0; 6], 0);
    nums1.iter().for_each(|&v| cnt[v as usize - 1] += 1);
    nums2.iter().for_each(|&v| cnt[6 - v as usize] += 1);
    for i in (1..=5).rev() {
      if d - i * cnt[i as usize] <= 0 {
        return ans + (d + i - 1) / i;
      }
      d -= cnt[i as usize] * i;
      ans += cnt[i as usize];
    }
    ans
  }
}
